src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs

```diff
@@ -113,21 +113,44 @@ public static partial class JsonSerializer
                     inputType = value!.GetType();
                 }
 
-                WriteStack state = default;
+                WriteStack state = new WriteStack { CancellationToken = cancellationToken };
                 JsonConverter converterBase = state.Initialize(inputType, options, supportContinuation: true);
 
                 bool isFinalBlock;
 
-                do
+                try
                 {
-                    state.FlushThreshold = (int)(bufferWriter.Capacity * FlushThreshold);
-
-                    isFinalBlock = WriteCore(converterBase, writer, value, options, ref state);
-
-                    await bufferWriter.WriteToStreamAsync(utf8Json, cancellationToken).ConfigureAwait(false);
-
-                    bufferWriter.Clear();
-                } while (!isFinalBlock);
+                    do
+                    {
+                        state.FlushThreshold = (int)(bufferWriter.Capacity * FlushThreshold);
+
+                        try
+                        {
+                            isFinalBlock = WriteCore(converterBase, writer, value, options, ref state);
+                        }
+                        finally
+                        {
+                            if (state.PendingAsyncDisposables?.Count > 0)
+                            {
+                                await state.DisposePendingAsyncDisposables().ConfigureAwait(false);
+                            }
+                        }
+
+                        await bufferWriter.WriteToStreamAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+                        bufferWriter.Clear();
+
+                        if (state.PendingTask is not null)
+                        {
+                            await state.AwaitPendingTask().ConfigureAwait(false);
```

Comment:
How likely are exceptions?  Seems like we're paying the cost to allocate another task here just to silence exceptions.  If we don't expect them to be likely, would it be better to just do:
```C#
if (state.PendingTask is not null)
{
    try
    {
        await state.PendingTask.ConfigureAwait(false);
    }
    catch
    {
        // Exceptions will be propagated elsewhere
        // TODO https://github.com/dotnet/runtime/issues/22144
    }
}
```
?

Reply:
Not unlikely, assuming the source IAE is IO bound. Agreed that a solution via #22144 is preferable.

Reply:
But, how common?  We're paying now for additional allocation each pending iteration, whereas with the above we'd pay for a throw/catch (which is more expensive) but only when something exceptional occurred.

Reply:
Agreed, and I've already made the change. Just pointing out that exceptions should be considered likely in this context.